### PR TITLE
[SPARK-20199][ML] : Provided featureSubsetStrategy to GBTClassifier and GBTRegressor

### PR DESCRIPTION
**Pull request by pralabhkumar**
## What changes were proposed in this pull request?

(Provided featureSubset Strategy to GBTClassifier
a) Moved featureSubsetStrategy to TreeEnsembleParams
b)  Changed GBTClassifier to pass featureSubsetStrategy
val firstTreeModel = firstTree.train(input, treeStrategy, featureSubsetStrategy))

## How was this patch tested?
a) Tested GradientBoostedTreeClassifierExample by adding .setFeatureSubsetStrategy with GBTClassifier

b)Added test cases in GBTClassifierSuite and GBTRegressorSuite

### CODE REVIEW

**Comment by pralabhkumar** [comment]
[~arushkharbanda][~peng.meng@intel.com][~facai] [~srowen]

Please review the pull request /approach,.

**Comment by pralabhkumar** [comment]
Can one of the admins please verify this patch?

**Comment by MLnick** [comment]
cc @sethah also

**Comment by MLnick** [comment]
ok to test

**Comment by pralabhkumar** [comment]
16ccbdf is successful . Please review the pull request .  
@MLnick @sethah @mpjlu @srowen

**Comment by pralabhkumar** [comment]
can any one of admin ,please review the pull request.  It would be really helpful. Thanks

File: mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala
Comment (mpjlu) [inline mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala]: delete blank line

File: mllib/src/main/scala/org/apache/spark/ml/regression/DecisionTreeRegressor.scala
Comment (mpjlu) [inline mllib/src/main/scala/org/apache/spark/ml/regression/DecisionTreeRegressor.scala]: delete this

File: mllib/src/main/scala/org/apache/spark/ml/tree/treeParams.scala
Comment (mpjlu) [inline mllib/src/main/scala/org/apache/spark/ml/tree/treeParams.scala]: two space

File: mllib/src/main/scala/org/apache/spark/ml/tree/treeParams.scala
Comment (mpjlu) [inline mllib/src/main/scala/org/apache/spark/ml/tree/treeParams.scala]: indent wrong

File: mllib/src/main/scala/org/apache/spark/ml/tree/treeParams.scala
Comment (mpjlu) [inline mllib/src/main/scala/org/apache/spark/ml/tree/treeParams.scala]: too many blank lines

File: project/MimaExcludes.scala
Comment (mpjlu) [inline project/MimaExcludes.scala]: is this the right place?

File: mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala
Comment (pralabhkumar) [inline mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala]: done

File: mllib/src/main/scala/org/apache/spark/ml/tree/treeParams.scala
Comment (pralabhkumar) [inline mllib/src/main/scala/org/apache/spark/ml/tree/treeParams.scala]: done

File: project/MimaExcludes.scala
Comment (pralabhkumar) [inline project/MimaExcludes.scala]: I put it in V21 excludes . Please let me know ,if you are expecting something else

**Comment by pralabhkumar** [comment]
@mpjlu Thanks for reviewing the code . I have done the code changes as suggested .
Build is passed with all test cases.

Please review and let me know if further changes are required.

**Comment by pralabhkumar** [comment]
@mpjlu : Please review the changes

**Comment by pralabhkumar** [comment]
12d83aa is successful . Please review the pull request .
@MLnick @sethah @mpjlu @srowen

**Comment by pralabhkumar** [comment]
can any one of admins review the pull request

**Comment by pralabhkumar** [comment]
can any one of admins please review the pull request

**Comment by pralabhkumar** [comment]
@jkbradley Please review the pull request

**Comment by pralabhkumar** [comment]
Have done the changes suggested by @mpjlu  . 

Please find some time to review the pull request .

**Comment by pralabhkumar** [comment]
@mpjlu 
please find some time to review the code.

**Comment by sethah** [comment]
I don't think there's any point in pinging every day :)

**Comment by pralabhkumar** [comment]
@sethah agree with you . Sorry if I unnecessary bother , was eager to get reviews on pull request. 
Thanks for the suggestion , will keep in mind

**Comment by pralabhkumar** [comment]
ping @mpjlu . Please review the pull request

**Comment by srowen** [comment]
@pralabhkumar if there is no support for this we should close it. Please stop commenting

**Comment by MLnick** [comment]
I think this may be a good feature to have, but haven't had bandwidth to
take a look and won't for the next week or so.

Not sure if @sethah has any time, or someone else who has been doing stuff
with trees recently.

As far as I can see this param setting is similar to xgboost "col sample by
level" - so it's not a feature sample by tree built (which is subsample
rate param in spark) but a feature sample by node built.

In any case RF does expose it and GBT doesn't and it does seem like it
could be exposed as an option. I don't see any downside or issue with it.

So it may just take some patience right now until someone can give a review.

**Comment by pralabhkumar** [comment]
@MLnick @srowen 
Thanks for your comments , I will wait for someone to review .    :)

File: mllib/src/main/scala/org/apache/spark/ml/regression/GBTRegressor.scala
Comment (sethah) [inline mllib/src/main/scala/org/apache/spark/ml/regression/GBTRegressor.scala]: use `$(featureSubsetStrategy)`

File: mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala
Comment (sethah) [inline mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala]: remove blank line

File: mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala
Comment (sethah) [inline mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala]: remove blank line

File: mllib/src/main/scala/org/apache/spark/ml/regression/DecisionTreeRegressor.scala
Comment (sethah) [inline mllib/src/main/scala/org/apache/spark/ml/regression/DecisionTreeRegressor.scala]: don't use a named parameter since the meaning is now clear

**Review by sethah — COMMENTED** [review]
Additionally:
* Change the title to obey the proper format `[SPARK-20199][ML] ...`
* Change title to reflect that both GBTClassifier and GBTRegressor are changed
* Please remove all the text you did not write from the PR description
* Add a test to check that the default values are correct for GBTClassifier/Regressor. See the test in logistic regression titled: "logistic regression: default params" for reference
* I'd like to test that this change takes effect. One way might be to construct a small dataset where one feature is highly predictive and other features are less so, train with `featureSubsetStrategy = "all"` and with `featureSubsetStrategy = "1"` and they should not produce the same tree. I'm open to other, simpler ways to test it if you can think of some.

File: mllib/src/main/scala/org/apache/spark/ml/tree/treeParams.scala
Comment (MLnick) [inline mllib/src/main/scala/org/apache/spark/ml/tree/treeParams.scala]: Why is this being moved around?

File: mllib/src/main/scala/org/apache/spark/ml/tree/treeParams.scala
Comment (MLnick) [inline mllib/src/main/scala/org/apache/spark/ml/tree/treeParams.scala]: Why is this being moved?

File: mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala
Comment (pralabhkumar) [inline mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala]: @sethah I tried to make it similar to RandomForest.scala . It have strategy and featureSubSetStrategy as separate parameter.

File: mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala
Comment (pralabhkumar) [inline mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala]: done

File: mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala
Comment (pralabhkumar) [inline mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala]: done

File: mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala
Comment (pralabhkumar) [inline mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala]: done

File: mllib/src/main/scala/org/apache/spark/mllib/tree/GradientBoostedTrees.scala
Comment (pralabhkumar) [inline mllib/src/main/scala/org/apache/spark/mllib/tree/GradientBoostedTrees.scala]: done

File: mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala
Comment (pralabhkumar) [inline mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala]: done

File: mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala
Comment (pralabhkumar) [inline mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala]: done . I will add this to GBTRegressor in next pull request (forgot to add in this one)

**Comment by pralabhkumar** [comment]
@sethah 

Thanks for reviewing the pull request .

- Change the title to obey the proper format [SPARK-20199][ML] ...

- Response : Done 

- Change  title to reflect that both GBTClassifier and GBTRegressor are changed

- Response : Done

- Please  remove all the text you did not write from the PR description

- Response : Done

- Add a test to check that the default values are correct for GBTClassifier/Regressor. See the test in logistic regression titled: "logistic regression: default params" for reference

- Response : Done

- I'd  like to test that this change takes effect. One way might be to construct a small dataset where one feature is highly predictive and other features are less so, train with featureSubsetStrategy = "all" and with featureSubsetStrategy = "1" and they should not produce the same tree. I'm open to other, simpler ways to test it if you can think of some.

- Response : Added test case to check for featureSubsetStrategy parameter. Creating two GBT trees ,one with subset strategy "all" and other with "1" . Comparing their most important feature and important features vector to make sure tree are different

File: mllib/src/main/scala/org/apache/spark/ml/tree/treeParams.scala
Comment (pralabhkumar) [inline mllib/src/main/scala/org/apache/spark/ml/tree/treeParams.scala]: @MLnick 
 
Earlier featureSubsetStrategy, setFeatureSubsetStrategy, getFeatureSubsetStrategy were part of RandomForestParams , moved them to be part of TreeEnsembleParams , so that it can be accessed by both Random forest and GBT . I have not moved anything apart from that .

**Comment by pralabhkumar** [comment]
@MLnick
Thanks for reviewing . I have added comment .  

Please  review them

File: mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala
Comment (pralabhkumar) [inline mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala]: @sethah 
Done

**Comment by sethah** [comment]
I'll take a look at the changes in the next few days. In the meantime, you can remove "Please review http://spark.apache.org/contributing.html before opening a pull request." from the PR description.

**Comment by pralabhkumar** [comment]
@sethah Thanks for comment . Have done the changes as suggested in PR description.  

I'll wait for the review comments from your side :)

**Comment by pralabhkumar** [comment]
ping @sethah , please let me know if there is any update on it . Thanks

**Comment by pralabhkumar** [comment]
ping @sethah @MLnick

File: mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala
Comment (MLnick) [inline mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala]: Not necessary to have an entirely new estimator here - you can just use ` val gbtWithFeatureSubset = gbt.setFeatureSubsetStrategy(...)`

File: mllib/src/test/scala/org/apache/spark/ml/regression/GBTRegressorSuite.scala
Comment (MLnick) [inline mllib/src/test/scala/org/apache/spark/ml/regression/GBTRegressorSuite.scala]: Same comment applies here

File: project/MimaExcludes.scala
Comment (MLnick) [inline project/MimaExcludes.scala]: Do we really still need this MiMa exclusion?

**Comment by pralabhkumar** [comment]
@MLnick Thanks for reviewing . Have done all the changes suggested by you . Please review .

File: mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala
Comment (pralabhkumar) [inline mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala]: done

File: mllib/src/test/scala/org/apache/spark/ml/regression/GBTRegressorSuite.scala
Comment (pralabhkumar) [inline mllib/src/test/scala/org/apache/spark/ml/regression/GBTRegressorSuite.scala]: done

File: project/MimaExcludes.scala
Comment (pralabhkumar) [inline project/MimaExcludes.scala]: @MLnick Yes you are correct , I have removed it . Thanks

**Comment by pralabhkumar** [comment]
ping @sethah  @MLnick

**Comment by pralabhkumar** [comment]
@MLnick @sethah ,please let me know ,if you are ok with the changes . So that we can proceed forward . Thanks for your help :)

**Comment by pralabhkumar** [comment]
ping @MLnick @sethah

**Comment by pralabhkumar** [comment]
@MLnick @sethah please find some time to look into this .It will be really great if we can include this feature in spark 2.3

**Comment by pralabhkumar** [comment]
@MLnick @sethah please find some time to look into this

**Comment by pralabhkumar** [comment]
@MLnick @sethah its been more than couple of months since the code changes has done as suggested. It would be really great if you can find some time to review it . Please review the pull request

**Comment by pralabhkumar** [comment]
@MLnick  @sethah  please find some time to look into this

File: mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala
Comment (sethah) [inline mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala]: Not sure if we need these tests. We already test the `featureSubsetStrategy` in the RandomForest implementation. At any rate, let's not set parameters that aren't strictly relevant to the test (stepSize, impurity, etc...).

File: mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala
Comment (sethah) [inline mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala]: the last two assertions here aren't necessary

File: mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala
Comment (sethah) [inline mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala]: last three assertions here aren't necessary. Also why not `!==`?

**Review by sethah — COMMENTED** [review]
Just a few minor things. Otherwise, looks good.

File: mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala
Comment (pralabhkumar) [reply mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala]: done

File: mllib/src/main/scala/org/apache/spark/ml/regression/DecisionTreeRegressor.scala
Comment (pralabhkumar) [reply mllib/src/main/scala/org/apache/spark/ml/regression/DecisionTreeRegressor.scala]: done

File: mllib/src/main/scala/org/apache/spark/ml/regression/DecisionTreeRegressor.scala
Comment (pralabhkumar) [reply mllib/src/main/scala/org/apache/spark/ml/regression/DecisionTreeRegressor.scala]: done

File: mllib/src/main/scala/org/apache/spark/ml/regression/DecisionTreeRegressor.scala
Comment (pralabhkumar) [reply mllib/src/main/scala/org/apache/spark/ml/regression/DecisionTreeRegressor.scala]: done

File: mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala
Comment (pralabhkumar) [reply mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala]: Removed stepSize , impurity and other  parameters

File: mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala
Comment (pralabhkumar) [reply mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala]: done

File: mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala
Comment (pralabhkumar) [reply mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala]: done

**Comment by pralabhkumar** [comment]
@sethah Thanks for reviewing code . I have done all the changed as suggested by you . 

Please review them and let me know if further changes are required.

**Comment by pralabhkumar** [comment]
@sethah please find some time to look into the changes .  

Please  let me know if further changes are required.

File: mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala
Comment (sethah) [reply mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala]: Setting maxIter and maxDepth to some small value is actually useful so the tests don't take unnecessarily long.

**Review by sethah — COMMENTED** [review]
One minor thing. Otherwise LGTM. Let's ping @MLnick or @jkbradley so we can move ahead with it.

**Comment by pralabhkumar** [comment]
@sethah Have done the changes as suggested ,but build is failing because of this error
Step ?Publish JUnit test result report? failed: No test report files were found. Configuration error?

https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83354/

Please help on the same

**Comment by sethah** [comment]
Jenkins test this please

**Comment by pralabhkumar** [comment]
@sethah Its still failing , I don't think so its issue from my side.  Please help

**Comment by pralabhkumar** [comment]
Jenkins test this please

**Comment by pralabhkumar** [comment]
@sethah Build is passed :) ,have done the changes as suggested (setting maxIter and maxDepth).

ping @MLnick or @jkbradley so we can move ahead with it.

**Comment by pralabhkumar** [comment]
@MLnick @jkbradley Please find some time to review it . @sethah has given LGTM

**Comment by pralabhkumar** [comment]
Ping @MLnick @jkbradley  . @sethah has given LGTM

File: mllib/src/main/scala/org/apache/spark/ml/regression/GBTRegressor.scala
Comment (MLnick) [inline mllib/src/main/scala/org/apache/spark/ml/regression/GBTRegressor.scala]: You removed this from the classifier but not here (as per https://github.com/apache/spark/pull/18118#discussion_r148063169)

File: mllib/src/test/scala/org/apache/spark/ml/regression/GBTRegressorSuite.scala
Comment (MLnick) [inline mllib/src/test/scala/org/apache/spark/ml/regression/GBTRegressorSuite.scala]: You've kept the other assertions in this test (that were removed from the classifier test as per https://github.com/apache/spark/pull/18118#discussion_r148095940). If they're not necessary we should also remove them here.

File: mllib/src/test/scala/org/apache/spark/ml/regression/GBTRegressorSuite.scala
Comment (MLnick) [inline mllib/src/test/scala/org/apache/spark/ml/regression/GBTRegressorSuite.scala]: Same here - we can remove unnecessary assertions as per https://github.com/apache/spark/pull/18118#discussion_r148096176

**Review by MLnick — CHANGES_REQUESTED** [review]
A few minor comments. Otherwise looks ready to go.

File: mllib/src/main/scala/org/apache/spark/ml/regression/GBTRegressor.scala
Comment (pralabhkumar) [reply mllib/src/main/scala/org/apache/spark/ml/regression/GBTRegressor.scala]: done

File: mllib/src/test/scala/org/apache/spark/ml/regression/GBTRegressorSuite.scala
Comment (pralabhkumar) [reply mllib/src/test/scala/org/apache/spark/ml/regression/GBTRegressorSuite.scala]: done

File: mllib/src/test/scala/org/apache/spark/ml/regression/GBTRegressorSuite.scala
Comment (pralabhkumar) [reply mllib/src/test/scala/org/apache/spark/ml/regression/GBTRegressorSuite.scala]: done

**Comment by pralabhkumar** [comment]
@MLnick  Thanks for the reviewing the code . Have done changes as suggested. 

Please proceed further if its good to go .

Thanks

**Comment by pralabhkumar** [comment]
@MLnick Please find some time to review it and let me know if we can proceed with this. Thanks

**Review by WeichenXu123 — APPROVED** [review]
LGTM. Good work!

**Comment by MLnick** [comment]
LGTM, merged to master. Thanks!

**Comment by pralabhkumar** [comment]
@MLnick @WeichenXu123 @sethah Thanks for your help throughout the process.

**Pull request closed without merging**
